### PR TITLE
Tweak: Added ability to opt-out AI features per user (#22796, #22853) [ED-15000] 

### PR DESCRIPTION
**Pull request by DennisNerush**

### CODE REVIEW

File: modules/ai/preferences.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		<h2><?php echo esc_html__( 'Elementor - AI', 'elementor' ); ?></h2>
+		<table class="form-table" role="presentation">
+			<tr>
+				<th>
+					<label for="<?php echo esc_attr( $ai_option_name ); ?>">
+						<?php echo esc_html__( 'Status', 'elementor' ); ?>
+					</label>
+				</th>
+				<td>
+					<label for="<?php echo esc_attr( $ai_option_name ); ?>">
```
Comment (TzviRabinovitch) [inline modules/ai/preferences.php:67]: Esc_attr is called multiple times

File: modules/ai/preferences.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		$wpnonce = Utils::get_super_global_value( $_POST, '_wpnonce' );
+		if ( ! wp_verify_nonce( $wpnonce, 'update-user_' . $user_id ) ) {
+			return;
+		}
+
+		if ( ! $this->has_permissions_to_edit_user( $user_id ) ) {
+			return;
+		}
+		
+		$ai_option_name = static::ENABLE_AI;
```
Comment (TzviRabinovitch) [inline modules/ai/preferences.php:96]: Why use local var ?

File: modules/ai/preferences.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	 * @param \WP_User $user - User object.
+	 *
+	 * @return void
+	 */
+	protected function add_personal_options_settings( \WP_User $user ) {
+		if ( ! $this->has_permissions_to_edit_user( $user->ID ) ) {
+			return;
+		}
+
+		$ai_option_name = static::ENABLE_AI;
```
Comment (TzviRabinovitch) [inline modules/ai/preferences.php:55]: Why use local var ?

File: modules/ai/preferences.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	 * Determine if the current user has permission to view/change preferences of a user.
+	 *
+	 * @param int $user_id
+	 *
+	 * @return bool
+	 */
+	protected function has_permissions_to_edit_user( $user_id ) {
+		return (
+			current_user_can( 'edit_user', $user_id )
+		);
```
Comment (louiswol94) [inline modules/ai/preferences.php:112]:
```suggestion
		return current_user_can( 'edit_user', $user_id );
```

File: modules/ai/preferences.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+	/**
+	 * Save the settings in the "Personal Options".
+	 *
+	 * @param int $user_id - User ID.
+	 *
+	 * @return void
+	 */
+	protected function update_personal_options_settings( $user_id ) {
+
```
Comment (louiswol94) [inline modules/ai/preferences.php:85]:
```suggestion
```
Redundant line

File: modules/ai/preferences.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	/**
+	 * Determine if AI features are enabled for a user.
+	 *
+	 * @param int $user_id - User ID.
+	 *
+	 * @return bool
+	 */
+	public static function is_ai_enabled( $user_id ) {
+		return ! ! User::get_user_option_with_default( static::ENABLE_AI, $user_id, true );
```
Comment (louiswol94) [inline modules/ai/preferences.php:40]:
```suggestion
		return (bool) User::get_user_option_with_default( static::ENABLE_AI, $user_id, true );
```
Maybe? More common and clear.

File: modules/ai/preferences.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		add_action( 'profile_personal_options', function ( \WP_User $user ) {
+			$this->add_personal_options_settings( $user );
+		} );
+
+		add_action( 'personal_options_update', function ( $user_id ) {
+			$this->update_personal_options_settings( $user_id );
+		} );
+	}
+
+
```
Comment (louiswol94) [inline modules/ai/preferences.php:30]:
```suggestion
```
Redundant lines

File: modules/ai/preferences.php [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
```
Comment (louiswol94) [inline modules/ai/preferences.php:31]:
```suggestion

```
Redundant lines

File: modules/ai/preferences.php (hunk 1 above)
Comment (DennisNerush) [reply modules/ai/preferences.php:67]: my php is a little rusty. Shouldn't I call it?

Comment (TzviRabinovitch) [reply modules/ai/preferences.php]: @DennisNerush 
Can you call it once and use it wherever it's needed ?

Comment (TzviRabinovitch) [reply modules/ai/preferences.php]: regardless that's ok with me , its a minor thing

File: modules/ai/preferences.php [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	const ENABLE_AI = 'elementor_enable_ai';
+
+	/**
+	 * Register actions and hooks.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_action( 'profile_personal_options', function ( \WP_User $user ) {
```
Comment (matipojo) [inline modules/ai/preferences.php:21]: Please replace with `personal_options` to show on every user option page

```suggestion
		add_action( 'personal_options', function ( \WP_User $user ) {
```

File: modules/ai/preferences.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+			$this->add_personal_options_settings( $user );
+		} );
+
+		add_action( 'personal_options_update', function ( $user_id ) {
+			$this->update_personal_options_settings( $user_id );
+		} );
```
Comment (matipojo) [inline modules/ai/preferences.php:27]:
```suggestion
		add_action( 'personal_options_update', function ( $user_id ) {
			$this->update_personal_options_settings( $user_id );
		} );
		
		add_action( 'edit_user_profile_update', function ( $user_id ) {
			$this->update_personal_options_settings( $user_id );
		} );
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/ai/preferences.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+	 * @param \WP_User $user - User object.
+	 *
+	 * @return void
+	 */
+	protected function add_personal_options_settings( \WP_User $user ) {
+		if ( ! $this->has_permissions_to_edit_user( $user->ID ) ) {
+			return;
+		}
+
+		$ai_value = User::get_user_option_with_default( static::ENABLE_AI, $user->ID, '1' );
```
Comment (matipojo) [inline modules/ai/preferences.php:53]: You can escape the id once and reuse:

```suggestion
		$ai_value = User::get_user_option_with_default( static::ENABLE_AI, $user->ID, '1' );
		
		$escaped_id = esc_attr( static::ENABLE_AI );
```

File: modules/announcements/module.php [hunk 12]
```diff
@@ -92,36 +93,42 @@ public function enqueue_styles() {
...
+					'variant' => 'secondary',
 				],
-				'triggers' => [
-					[
-						'action' => 'aiStared',
-					],
+			],
+			'triggers' => [
+				[
+					'action' => 'aiStared',
```
Comment (matipojo) [inline modules/announcements/module.php:128]:
```suggestion
					'action' => 'aiStarted',
```

**Review by matipojo — DISMISSED** [review]

**Review by matipojo — APPROVED** [review]

**Pull request merged**